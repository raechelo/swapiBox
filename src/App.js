import React, { Component } from 'react';
import { Route, NavLink, Link } from 'react-router-dom';
import People from './People/People';
import Loader from './Loader/Loader'
import CardContainer from './CardContainer/CardContainer';
import Starfield from './Starfield/Starfield'
import Crawl from './Crawl/Crawl'

class App extends Component {
  constructor() {
    super();
    this.state = {
      people: [],
      isLoading: true,
      movie: {}
    }
  }

  fetchFilms = () => {
    const movieUrl = 'https://swapi.co/api/films/';
    fetch(movieUrl)
      .then(response => response.json())
      .then(data => data.results.sort(() => 0.5 - Math.random()).pop()  )
      .then(movie => this.setState( {movie} ) )
  }

  componentDidMount() {
    const peopleUrl = 'https://swapi.co/api/people/';
    fetch(peopleUrl)
      .then(response => response.json())
      .then(data => this.setState( { people: data.results, isLoading: false } ) )
      .then(data => this.fetchFilms())
      .catch(err => console.log(err))
  }



  render() {

    return (
      <div className="App">
      <Starfield className="Starfield" />
        <h1 className="header">swapiboxi</h1>
        <nav>
          <button>People</button>
          <button>Planets</button>
          <button>Vehicles</button>
          <button>Favorites</button>
        </nav>
        {
          this.state.isLoading === true ? 
          <Loader /> : 
          <Crawl 
          crawl={this.state.movie.opening_crawl} 
          release={this.state.movie.release_date} 
          />
        }
      </div>
    );
  }
}

export default App;
