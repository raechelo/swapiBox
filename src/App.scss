.App {
  color: $accent;
  display: grid;
  grid-template-columns: 20% 80%;
  grid-template-areas: "title title"
  "btns cards";
  font-family: $fancyfont;
  text-align: center;
}

.header {
  grid-area: title;
}

nav {
  z-index: 15;
}

nav button {
  background-color: transparent;
  border: 3px solid $accent;
  color: $accent;
  font-family: $fancyfont;
  font-size: 1.5rem;
  grid-area: btns;;
  display: block;
  margin: 2rem auto;
  padding: 5px;
  width: 80%;
  &:hover {
    background-color: $accent;
    color: black;
    cursor: pointer;
  }
}

.movie-scroll {
  color: $accent;
}