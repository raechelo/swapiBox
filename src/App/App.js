import React, { Component } from 'react';
import Loader from '../Loader/Loader'
import CardContainer from '../CardContainer/CardContainer';
import Starfield from '../Starfield/Starfield'
import propTypes from 'prop-types';

class App extends Component {
  constructor() {
    super();
    this.state = {
      isLoading: true,
      movie: {},
      people: [],
      planets: [], 
      vehicles: [],
      favorites: [],
      currentChoice: '',
    }
  }

  fetchPeople = () => {
    const peopleUrl = 'https://swapi.co/api/people/';
    fetch(peopleUrl)
      .then(response => response.json())
      .then(data => this.setState( { people: data.results, isLoading: false, currentChoice: 'crawl' } ) )
      .then(() => this.fetchHomeworlds(this.state.people))
      .catch(err => console.log(err));
  }

  fetchHomeworlds = (arr) => {
    let homeworlds = arr.map(p => {
      fetch(p.homeworld)
        .then(response => response.json())
        .then(data => this.addHomeworldInfo(data.name, data.population) )
        .catch(err => console.log(err))
    })
    this.fetchSpecies(this.state.people);
    return Promise.all(homeworlds)
  }

  addHomeworldInfo = (name, pop) => {
    const addInfo = {homeworld:name, homeworldPopulation: pop}
    const people = this.state.people.map(p => {
      return Object.assign(p, addInfo)
    })
    this.setState( { people } )
  }

  fetchSpecies = (arr) => {
    let species = arr.map(p => {
      fetch(p.species)
        .then(response => response.json())
        .then(data => this.addSpeciesInfo(data.name))
        .catch(err => console.log(err))
    })
    return Promise.all(species)
  }

  addSpeciesInfo = (species) => {
    const addSpecies = {species:species}
    const people = this.state.people.map(p => {
      return Object.assign(p, addSpecies)
    })
    this.setState( {people } );
  }

  componentDidMount() {
    const movieUrl = 'https://swapi.co/api/films/';
    fetch(movieUrl)
      .then(response => response.json())
      .then(data => data.results.sort(() => 0.5 - Math.random()).pop()  )
      .then(movie => this.setState( { movie } ) )
    this.fetchPeople();
    this.fetchVehicles();
    this.fetchPlanets();
  }

  fetchVehicles = () => {
    const vehicleUrl = 'https://swapi.co/api/vehicles/';
    fetch(vehicleUrl)
      .then(response => response.json())
      .then(data => this.setState( {vehicles: data.results} ) )
      .catch(err => console.log(err))
  }

  fetchPlanets = () => {
    const planetUrl = 'https://swapi.co/api/planets/';

    fetch(planetUrl)
      .then(response => response.json())
      .then(data => this.setState( { planets: data.results } ) )
      .catch(err => console.log(err))
  }

  handleClick = (e) => {
    switch(e.target.textContent) {
      case 'people' :
        this.setState( { currentChoice: 'people' } )
      break;
      case 'vehicles':
        this.setState( { currentChoice: 'vehicles' } )
      break;
      case 'planets':
        this.setState( { currentChoice: 'planets' } )
      break;
      case 'favorites':
        this.setState( { currentChoice: 'favorites' } )
      break;
      default:
        this.setState( { currentChoice: 'crawl' } )
    }
  }


  //give button own component and do an isClicked state with ternary & inline styling to denote active state buttons

  render() {

    return (
      <div className="App">
      <Starfield className="Starfield" />
        <h1 className="header"><img src="https://fontmeme.com/permalink/190502/5dfed28f102f9d81a68ccee2d51fa26a.png" alt="swapiboxi" border="0" />
        </h1>
        <nav>
          <button className="people" onClick={this.handleClick} >people</button>
          <button className="planets" onClick={this.handleClick} >planets</button>
          <button className="vehicles" onClick={this.handleClick} >vehicles</button>
          <button className="favorites" onClick={this.handleClick} >favorites<span className="fave-count">{this.state.favorites.length}</span></button>
        </nav>
        {
          this.state.isLoading === true ? 
          <Loader /> : 
          <CardContainer rendered={this.state} />
        }
      </div>
    );
  }
}



App.protoTypes = {
  people: propTypes.array,
  isLoading: propTypes.bool,
  movie: propTypes.objct,
  planets: propTypes.array,
  vehicles: propTypes.array,
  favorites: propTypes.array,
  currentChoice: propTypes.string
}


export default App;