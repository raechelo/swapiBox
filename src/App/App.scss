.App {
  color: $accent;
  display: grid;
  grid-template-columns: 20% 80%;
  grid-template-areas: "title title"
  "btns cards";
  font-family: $fancyfont;
  text-align: center;
}

.header {
  grid-area: title;
}

nav {
  border-right: 2px solid $accent;
  z-index: 15;
}

nav button {
  background-color: hsla(51, 100%, 50%, 0.28);
  border: 3px solid $accent;
  color: $accent;
  font-family: $fancyfont;
  font-size: 1.5rem;
  grid-area: btns;;
  display: block;
  margin: 2rem auto;
  padding: 5px;
  text-transform: uppercase;
  width: 80%;
  &:hover, &:focus {
    background-color: $accent;
    color: black;
    cursor: pointer;
  }
}

.fave-count {
  margin: 0 .5rem;
  text-decoration: underline;
}

.active {
  background-color: $accent;
  color: black;
  cursor: pointer;
}

.crawl-container {
  padding: 1rem 2rem;
}