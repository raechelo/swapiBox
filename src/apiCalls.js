const fetchCalls = url => {
  return fetch (url, {
      method: 'GET'

  })
  .then(response => {
    if (!response.ok) {
      throw Error('Error fetching data')
    } else {
      return response.json()
    }
  })
}

export { fetchCalls }