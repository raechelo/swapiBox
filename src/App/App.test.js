import React from 'react';
import App from './App';
import { shallow } from 'enzyme';
import { fetchCalls } from '../apiCalls';

describe('App', () => {


  let wrapper;
  let mockFetchPlanets = jest.fn();
  let mockFetchPeople = jest.fn();
  let mockFetchVehicles = jest.fn();
  let mockAddPlanetInfo = jest.fn();

  let peopleArr = [
        {
            "name": "Luke Skywalker",
            "height": "172",
            "mass": "77",
            "hair_color": "blond",
            "skin_color": "fair",
            "eye_color": "blue",
            "birth_year": "19BBY",
            "gender": "male",
            "homeworld": "https://swapi.dev/api/planets/1/",
            "films": [
                "https://swapi.dev/api/films/2/",
                "https://swapi.dev/api/films/6/",
                "https://swapi.dev/api/films/3/",
                "https://swapi.dev/api/films/1/",
                "https://swapi.dev/api/films/7/"
            ],
            "species": [
                "https://swapi.dev/api/species/1/"
            ],
            "vehicles": [
                "https://swapi.dev/api/vehicles/14/",
                "https://swapi.dev/api/vehicles/30/"
            ],
            "starships": [
                "https://swapi.dev/api/starships/12/",
                "https://swapi.dev/api/starships/22/"
            ],
            "created": "2014-12-09T13:50:51.644000Z",
            "edited": "2014-12-20T21:17:56.891000Z",
            "url": "https://swapi.dev/api/people/1/"
        }
    ]

  beforeEach(() => {
    wrapper = shallow(
      <App />
    )

    window.fetch = jest.fn().mockImplementation(() => {
      return Promise.resolve({
        ok: true,
        status: 200,
        json: () => Promise.resolve()
      })
    })
  })

  it('should match the snapshot with all the data passed in', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have proper default state', () => {
    expect(wrapper.state()).toEqual( { isLoading: true, movie: {}, people: [], planets: [], vehicles: [], favorites: [], currentChoice: '' } )
  });

  
  it('should pull in the data with the correct movie url', () => {
    const expected = [ 'https://swapi.dev/api/films/', { method: 'GET' } ];
    const wrapper = shallow(<App />);
    
    expect(window.fetch).toHaveBeenCalledWith(...expected);
    
    // expect(mockFetchPeople).toHaveBeenCalledTimes(1);
    // expect(mockFetchVehicles).toHaveBeenCalledTimes(1);
    // expect(mockFetchPlanets).toHaveBeenCalledTimes(1);
  });

  it('should handle any clicks', () => {
    expect(wrapper.state('currentChoice')).toEqual('');

    wrapper.instance().handleClick( { target: { textContent: 'people' } } );
    expect(wrapper.state('currentChoice')).toEqual('people');

    wrapper.instance().handleClick( { target: { textContent: 'planets' } } );
    expect(wrapper.state('currentChoice')).toEqual('planets');

    wrapper.instance().handleClick( { target: { textContent: 'vehicles' } } );
    expect(wrapper.state('currentChoice')).toEqual('vehicles');

    wrapper.instance().handleClick( { target: { textContent: 'favorites' } } );
    expect(wrapper.state('currentChoice')).toEqual('favorites');
  });

  it('should pull in the data with the correct people url', async () => {
    const expected = ['https://swapi.dev/api/people/', {
      method: 'GET'
    }];

    wrapper.instance().fetchPeople()
    expect(window.fetch).toHaveBeenCalledWith(...expected);
    expect(window.fetch).toHaveBeenCalledTimes(1);
    expect(wrapper.state('people')).toEqual( [] )
  });

  it('should pull in the data with the correct homeworld name and population', () => {
    const expected = ['https://swapi.dev/api/planets/1/', { method: 'GET' } ];

    wrapper.setState({people: peopleArr})
    wrapper.instance().fetchHomeworlds(peopleArr);
    expect(window.fetch).toHaveBeenCalledWith(...expected)
  });

  it('should combine homeworld information', () => {
    const person = [ {name: 'Thanos' } ];
    
    wrapper.setState( { people: person } )
    wrapper.instance().addHomeworldInfo('Titan', 9)
    expect(wrapper.state('people')).toEqual( [ { name:'Thanos', homeworld: 'Titan', homeworldPopulation: 9 } ] ) 
  })

  it.skip('should set state after pulling in correct data', async () => {
    expect(wrapper.state('people')).toEqual([]);

    await wrapper.instance().fetchPeople();
    
    expect(wrapper.state('isLoading')).toEqual(false);
    expect(wrapper.state('currentChoice')).toEqual('crawl');
    expect(wrapper.state('people')).toEqual(peopleArr)
  })

  it('should pull in the data with the correct species name', () => {
    const expected = [['https://swapi.dev/api/species/1/'], { method: 'GET' } ];

    wrapper.instance().fetchSpecies(peopleArr);
    expect(window.fetch).toHaveBeenCalledWith(...expected);
  });

  it('should combine any species info', () => {
    const person = [ {name: 'Thanos' } ];

    wrapper.setState( { people: person } );
    wrapper.instance().addSpeciesInfo('Titan');
    expect(wrapper.state('people')).toEqual( [ { name: 'Thanos', species: 'Titan' } ] )
  });

  it('should pull in all correct data for the vehicles', async () => {
    const expected = ['https://swapi.dev/api/vehicles/', { method: 'GET' } ]

    wrapper.instance().fetchVehicles();
    expect(window.fetch).toHaveBeenCalledWith(...expected);
  });

  it('should combine the planet information', () => {
    const planet = [ { name: 'Mordor' } ];
    let info = ['Sauron'];
    
    wrapper.setState( { planets: planet } );
    wrapper.instance().addPlanetInfo([info]);
    expect(wrapper.state('planets')).toEqual( [ { name: 'Mordor', residents: info } ] );
  });

  it('should fetch any residents', () => {
    const planet = [ { name: 'Mordor', residents: ['https://swapi.dev/api/people/1/'] } ];
    const expected = ['https://swapi.dev/api/people/1/', { method: 'GET' } ];

    wrapper.setState( { planets: planet } )    
    wrapper.instance().fetchResidents(planet);
    expect(window.fetch).toHaveBeenCalledWith(...expected);
    expect(wrapper.state('planets')).toEqual( [ { name: 'Mordor', residents: [] } ] );
    // expect(mockAddPlanetInfo).toHaveBeenCalledTimes(1);
  })

  it('should pull in the planet data, including resident links', () => {
    const expected = ['https://swapi.dev/api/planets/', { method: 'GET' } ]

    wrapper.instance().fetchPlanets()
    expect(window.fetch).toHaveBeenCalledWith(...expected);
  });

  it('should favorite any items', () => {
    const person = { name:'Buzz Lightyear' }
    
    wrapper.instance().favoriteItem(person);
    expect(wrapper.state('favorites')).toEqual([person])
  });

  it('should remove items from favorites', () => {
    const person = [ { name:'Buzz Lightyear' } ];

    wrapper.setState({favorites: person});
    wrapper.instance().removeFavorites('Buzz Lightyear');
    expect(wrapper.state('favorites')).toEqual([])
  })

  it('should return an error if fetch does not return', async () => {
    window.fetch = jest.fn().mockImplementation(() => {
      return Promise.resolve({
        ok: false
      });
    })
    await expect(fetchCalls()).rejects.toEqual(Error('Error fetching data'))
  });

})

