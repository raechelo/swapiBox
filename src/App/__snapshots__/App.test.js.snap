// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App should match the snapshot with all the data passed in 1`] = `
<div
  className="App"
>
  <Starfield
    className="Starfield"
  />
  <h1
    className="header"
  >
    <img
      alt="swapiboxi"
      border="0"
      src="https://fontmeme.com/permalink/190502/5dfed28f102f9d81a68ccee2d51fa26a.png"
    />
  </h1>
  <nav>
    <button
      className="people"
      onClick={[Function]}
    >
      people
    </button>
    <button
      className="planets"
      onClick={[Function]}
    >
      planets
    </button>
    <button
      className="vehicles"
      onClick={[Function]}
    >
      vehicles
    </button>
    <button
      className="favorites"
      onClick={[Function]}
    >
      favorites
      <span
        className="fave-count"
      >
        0
      </span>
    </button>
  </nav>
  <Loader />
</div>
`;
